### ISSUE TITLE
V7.7 Error 500 with UTF8 & Special Characters!

### ISSUE THREAD

**Issue by exuperok** [gitlabhq/gitlabhq#8709]
1. **Summary:** Gitlab 7.7.2 ActionView component returns an Error 500 irelated to UTF8 encoding whenever it encounters a special character like  'ç'.
 
2. **Steps to reproduce:**
This issue happens when trying to perform the following actions
 
    * Displaying the file tree of a Branch 
    * Displaying the commit list of the branch 
    * Displaying the network of a branch  
    * Displaying the the graph of a branch 
    * Displaying the compare new merge request page  

3. **Displaying File Tree:**
     * **Expected behavior:**  displying the file tree of a branch including commits data.
     * **Observed behavior:** The commit data column keep displaying "loading data" message

     * **Relevant logs:**
```Processing by Projects::RefsController#logs_tree as JS 
  Parameters: {"_"=>"1422632454051", "project_id"=>"apis/api-exposure-layer", "id"=>"master"} 
Completed 500 Internal Server Error in 325ms 
ActionView::Template::Error (invalid byte sequence in UTF-8): 
    1: %span.str-truncated 
    2:   %span.tree_author= commit_author_link(commit, avatar: true, size: 16) 
    3:   = link_to_gfm commit.title, project_commit_path(@project, commit.id), class: "tree-commit-link" 
This seems to suggest that the (invalid byte sequence in UTF-8) is located```


4. **Displaying compare branches in merge request:**

     * **Expected behavior:**  displying the page that compares branches involved in merge request
     * **Observed behavior:** A blank pange with the error 500 is displayed

     * **Relevant logs:**
```Completed 500 Internal Server Error in 220ms 
ActionView::Template::Error (invalid byte sequence in UTF-8): 
    26:         = preserve(gfm(escape_once(commit.description))) 
    27:  
    28:   .commit-row-info 
    29:     = commit_author_link(commit, avatar: true, size: 16) 
    30:     .committed_ago 
    31:       #{time_ago_with_tooltip(commit.committed_date)}```


5. **Displaying the commit list of the branch:**

     * **Expected behavior:**  displying the list of commits of a branch
     * **Observed behavior:** The commit data column keep displaying "loading data" message

     * **Relevant logs:**
*Completed 500 Internal Server Error in 233ms 
```ActionView::Template::Error (invalid byte sequence in UTF-8): 
    26:         = preserve(gfm(escape_once(commit.description))) 
    27:  
    28:   .commit-row-info 
    29:     = commit_author_link(commit, avatar: true, size: 16) 
    30:     .committed_ago 
    31:       #{time_ago_with_tooltip(commit.committed_date)}```


6. **Displaying the network of a branch:**

     * **Expected behavior:**  displying thenetwork of a branch
     * **Observed behavior:** blank page

     * **Relevant logs:**

```Completed 500 Internal Server Error in 2003ms 
ActionView::Template::Error (source sequence is illegal/malformed utf-8): 
  app/views/projects/network/show.json.erb:22:in `_app_views_projects_network_show_json_erb__4224933065517375966_111137440'```


7. **Displaying the graph of a branch:**

     * **Expected behavior:**  displying the graph of a branch 
     * **Observed behavior:** blank page

     * **Relevant logs and/or screenshots:**
```Completed 500 Internal Server Error in 163ms 
JSON::GeneratorError (source sequence is illegal/malformed utf-8): 
  app/controllers/projects/graphs_controller.rb:37:in `fetch_graph' 
  app/controllers/projects/graphs_controller.rb:10:in `block (2 levels) in show' 
  app/controllers/projects/graphs_controller.rb:7:in `show'```

1. **Output of checks**
    *  **GitLab Version:** 7.7.2

    *  **Last commit SHA-1:** bd98290

    * **System Information:**
       ```System:         CentOS 6.5
           Current User:   git
           Using RVM:      no
           Ruby Version:   2.1.4p265
           Gem Version:    2.2.1
           Bundler Version:1.5.3
           Rake Version:   10.3.2
           Sidekiq Version:2.17.8```

    * **Results of GitLab [Application Check]:**

```Checking Environment ...
Git configured for git user? ... yes
Checking Environment ... Finished
Checking GitLab Shell ... 
GitLab Shell version >= 2.4.2 ? ... OK (2.4.2)
Repo base directory exists? ... yes
Repo base directory is a symlink? ... no
Repo base owned by git:git? ... yes
Repo base access is drwxrws---? ... yes
Satellites access is drwxr-x---? ... yes
hooks directories in repos are links: ... 
3/1 ... repository is empty
3/3 ... ok
Running /opt/gitlab/embedded/service/gitlab-shell/bin/check
Check GitLab API access: OK
Check directories and files: 
        /var/opt/gitlab/git-data/repositories: OK
        /var/opt/gitlab/.ssh/authorized_keys: OK
Test redis-cli executable: redis-cli 2.8.2
Send ping to redis server: PONG
gitlab-shell self-check successful
Checking GitLab Shell ... Finished
Checking Sidekiq ...
Running? ... yes
Number of Sidekiq processes ... 1
Checking Sidekiq ... Finished
Checking LDAP ...
LDAP is disabled in config/gitlab.yml
Checking LDAP ... Finished
Checking GitLab ...
Database config exists? ... yes
Database is SQLite ... no
All migrations up? ... yes
Database contains orphaned GroupMembers? ... no
GitLab config exists? ... yes
GitLab config outdated? ... no
Log directory writable? ... yes
Tmp directory writable? ... yes
Init script exists? ... skipped (omnibus-gitlab has no init script)
Init script up-to-date? ... skipped (omnibus-gitlab has no init script)
projects have namespace: ... 
3/1 ... yes
3/3 ... yes
Projects have satellites? ... 
3/1 ... can't create, repository is empty
3/3 ... yes
Redis version >= 2.0.0? ... yes
Ruby version >= 2.0.0 ? ... yes (2.1.4)
Your git bin path is /opt/gitlab/embedded/bin/git
Git version >= 1.7.10 ? ... yes (2.0.5)
Checking GitLab ... Finished ```

**Comment by dosire** [comment]
@DouweM there are a lot of issues on the issue tracker about using special characters in branches with GitLab, can you see if you can reproduce on the latests version of GitLab?

**Comment by exuperok** [comment]
In this case the incriminated special character was the "ç" which is part of the name of one the commiter in my team.

**Comment by exuperok** [comment]
tried this in gitlab.rb with no sucess.
# gitlab_rails['env'] = {
 #   'BUNDLE_GEMFILE' => "/opt/gitlab/embedded/service/gitlab-rails/Gemfile",
 #   'PATH' => "/opt/gitlab/bin:/opt/gitlab/embedded/bin:/bin:/usr/bin",
 #   'Encoding.default_internal' => "UTF-8",
 #   'Encoding.default_external' => "UTF-8"
 # }
having difficulty setting an environment for development to research a patch with an omnibus install. Is there a guide somewhere on how to setup a dev environment for gitlab so i can search a patch and submit for approval?

**Comment by exuperok** [comment]
@dosire @DouweM you can reproduce those issues on this private repository on gitlab.com. regards

**Comment by DouweM** [comment]
@exuperok Thanks, I can reproduce the issue on gitlab.com. I'll clone and investigate the issue locally.

**Comment by DouweM** [comment]
@exuperok All right, I've got an idea of how to solve this problem once and for all. Working on a fix.

**Comment by exuperok** [comment]
Yes by the after applying a hugly patch to mke it work in our entreprise, i just realised that i get this when calling api/v3 JSON::GeneratorError (source sequence is illegal/malformed utf-8):
  /opt/gitlab/embedded/service/gem/ruby/2.1.0/gems/activesupport-4.1.1/lib/active_support/core_ext/object/json.rb:34:in `to_json'

this happens for the following routes

GET "/api/v3/projects/projects:id/repository/tags" 
GET "/api/v3/projects/projects:id/repository/branches"

**Comment by exuperok** [comment]
by the way thanks for you attention and good luck in working on th fix, hopefully it'll see the light of the day very soon.

**Comment by jmeiss** [comment]
I'm experiencing the same issue. I suppose @exuperok you are french? ;)

I solved it by removing special char but it's more a workaround than a fix.

Thanks for your help!

**Comment by bpineau** [comment]
Same here. Also happens when creating merge requests with branchs containing non-utf-8 commits (non-utf-8 on Author or message field), and on /api/v3/projects/:id/repository/commits?ref_name=... routes.

**Comment by DouweM** [comment]
We're working on a fix to all of those issues, no need to +1 anymore ;)

**Comment by DouweM** [comment]
A fix to this issue has been merged and will be in version 7.9. Please reopen this issue if the problem persists after upgrading to 7.9.
